src/metrics_clustering.rs

```diff
@@ -0,0 +1,221 @@
+    /// silhouette score of each sample.
+    fn silhouette_score(&'a self) -> Result<F>;
+}
+
+struct DistanceCount<F> {
+    total_distance: F,
+    count: usize,
+}
+
+impl<F: Float> DistanceCount<F> {
+    /// Sets the total distance from the sample to this cluster to zero
+    pub fn reset(&mut self) {
+        self.total_distance = F::zero();
+    }
+
+    pub fn new(count: usize) -> DistanceCount<F> {
+        DistanceCount {
+            total_distance: F::zero(),
+            count,
+        }
+    }
+
+    /// Divides the total distance from the sample to this cluster by the number of samples in the cluster
+    pub fn mean_distance(&self) -> F {
+        self.total_distance / F::from(self.count).unwrap()
+    }
+
+    /// To be used in the cluster in which the sample is located. The distance from the sample to itself
+    /// is zero so it does not get added to the total distance. We can then just divide the total
+    /// distance by 1 - #samples in this cluster
+    pub fn same_label_mean_distance(&self) -> F {
+        if self.count == 1 {
+            return F::zero();
+        }
+        self.total_distance / F::from(self.count - 1).unwrap()
+    }
+
+    /// adds the distance of `other_sample` from `eval_sample` to the total distance of `eval_sample` from the current cluster
+    pub fn add_point(&mut self, eval_sample: ArrayView1<F>, other_sample: ArrayView1<F>) {
+        self.total_distance += F::from(eval_sample.l2_dist(&other_sample).unwrap()).unwrap();
+    }
+}
+
+impl<
+        'a,
+        F: Float,
+        L: 'a + Label,
+        D: Data<Elem = F>,
+        T: AsTargets<Elem = L> + FromTargetArray<'a, L> + Labels<Elem = L>,
+    > SilhouetteScore<'a, F> for DatasetBase<ArrayBase<D, Ix2>, T>
+{
+    fn silhouette_score(&'a self) -> Result<F> {
+        // By using try_single_target we ensure that the iterator returns an
+        // array1 as target with just one element, that can be addressed by [0]
+        let mut labels: HashMap<L, DistanceCount<F>> =
+            self.targets().try_single_target()?.label_count()[0]
+                .to_owned()
+                .into_iter()
+                .map(|(label, count)| (label, DistanceCount::new(count)))
+                .collect();
+
+        // Single label dataset, all points are in the same cluster.
+        if labels.len() == 1 {
+            return Ok(F::one());
+        }
+
+        // Compute and sum silhouette score for each sample
+        let score = self
+            .sample_iter()
+            .map(|sample| {
+                // Loops through all samples in the dataset and adds
+                // the distance between them and `sample` to the cluster
+                // in which they belong
+
+                for other in self.sample_iter() {
+                    labels
+                        .get_mut(&other.1[0])
+                        .unwrap()
+                        .add_point(sample.0, other.0);
```

Comment:
With `sample` fixed in the other loop, `add_point` should be called for every possible `other`, because I need to compute the distance between `sample` and every other point in the dataset

Reply:
but isn't the point of the silhouette score to compare the mean distance to min distance within a single cluster and not to all data points in the dataset?

Reply:
Yes but I need to find which cluster is the closest (on average), and for that, I need to compute the mean distance between `sample` and the points in each cluster

Reply:
ah I see, so you have look at each combination anyways! I was just confused, because you could add distances multiple times, but then saw the resetting operation in the next loop :+1: